### PR TITLE
Rework the itemCreate service

### PR DESCRIPTION
**Pull request by zenovich**
[itemCreate](https://docs.google.com/document/d/15Xxmatrs_7my5ioriikL6YgOU4I0sAVcq5ySKinc0Nw/edit#heading=h.uw1uftbl7kwt)

### CODE REVIEW

File: app/api/items/add_item.feature [hunk 1]
```diff
@@ -2,74 +2,209 @@ Feature: Add item
...
+          "access_open_date": "2018-01-02T03:04:05Z",
+          "duration": "01:02:03",
+          "end_contest_date": "2019-02-03T04:05:06Z",
+          "show_user_infos": true,
+          "contest_phase": "Analysis",
+          "level": 345,
+          "no_score": true,
+          "group_code_enter": true
+        },
+        "language_id": "3",
```
Comment (smadbe) [inline app/api/items/add_item.feature]: this format is bit strange with "item", "language_id", "string" at the root. It looks like something constraints by technical reasons ;-)

File: app/api/items/add_item.go [hunk 2]
```diff
@@ -3,145 +3,292 @@ package items
...
-func (in *NewItemRequest) groupItemData(id, userID, groupID int64) *database.GroupItem {
-	return &database.GroupItem{
-		ID:             *types.NewInt64(id),
-		ItemID:         in.ID.Int64,
-		GroupID:        *types.NewInt64(groupID),
-		CreatorUserID:  *types.NewInt64(userID),
-		FullAccessDate: *types.NewDatetime(time.Now()),
-		OwnerAccess:    *types.NewBool(true),
-		ManagerAccess:  *types.NewBool(true),
+func (in *NewItemRequest) groupItemData(groupItemID, userID, groupID, itemID int64) map[string]interface{} {
```
Comment (smadbe) [inline app/api/items/add_item.go]: When you read these functions alone, it is not easy to understand what it stands for. It is clearer when you see their call. You should add a simple comment such as "Creating the map containing the db data to be inserted into the GroupItem table" etc

File: app/api/items/add_item.go [hunk 3]
```diff
@@ -3,145 +3,292 @@ package items
...
+
+		userSelfGroupID, _ := user.SelfGroupID() // the user has been already loaded in addItem()
+		service.MustNotBeError(s.GroupItems().InsertMap(newItemRequest.groupItemData(store.NewID(), user.UserID, userSelfGroupID, itemID)))
+		service.MustNotBeError(s.ItemStrings().InsertMap(newItemRequest.itemStringData(store.NewID(), itemID, stringMap)))
+
+		service.MustNotBeError(
+			s.ItemItems().InsertMap(newItemRequest.itemItemData(store.NewID(), newItemRequest.ParentItemID, itemID, newItemRequest.Order)))
+
+		for _, child := range newItemRequest.Children {
+			service.MustNotBeError(s.ItemItems().InsertMap(newItemRequest.itemItemData(store.NewID(), itemID, child.ItemID, child.Order)))
```
Comment (smadbe) [inline app/api/items/add_item.go]: isn't there multi-insert in MySQL & gorm?

File: app/api/items/add_item.robustness.feature [hunk 4]
```diff
@@ -1,211 +1,953 @@
...
+        },
+        "parent_item_id": "21",
+        "order": 100,
+        "children": [
+          {"item_id": "4", "order": 1}
+        ]
+      }
+      """
+    Then the response code should be 403
+    And the response error message should contain "An item cannot become an ancestor of itself"
```
Comment (smadbe) [inline app/api/items/add_item.robustness.feature]: Wow, kudo for these 1000 lines of scenarios!  🤗

File: app/database/item_store.go [hunk 5]
```diff
@@ -79,26 +79,21 @@ func (s *ItemStore) Insert(data *Item) error {
...
-		Scan(&dbRes)
-	if db.Error() != nil {
-		return false, false, db.Error()
-	}
-	if len(dbRes) != 1 {
-		return false, false, nil
+	err = s.GroupItems().MatchingUserAncestors(user).
+		WithWriteLock().
+		Where("idItem IN (?) AND (bManagerAccess OR bOwnerAccess)", itemIDs).
+		PluckFirst("COUNT(DISTINCT idItem)", &count).Error()
```
Comment (smadbe) [inline app/database/item_store.go]: nice one 👍 
Shouldn't you write a `count(..)` fct on db to make it more readable?

File: app/database/db.go [hunk 6]
```diff
@@ -455,6 +456,37 @@ func (conn *DB) insert(tableName string, data interface{}) error {
...
+	var values = make([]interface{}, 0, len(dataMap))
+
+	keys := make([]string, 0, len(dataMap))
+	for key := range dataMap {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	for _, key := range keys {
+		attributes = append(attributes, key)
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/database/db.go]: fixed

File: app/api/items/add_item.go [hunk 7]
```diff
@@ -3,145 +3,292 @@ package items
...
-		ID:                in.ID.Int64,
-		Type:              in.Type.String,
-		DefaultLanguageID: in.Strings[0].LanguageID.Int64,
-		TeamsEditable:     *types.NewBool(false), // has no db default at the moment, so must be set
-		NoScore:           *types.NewBool(false), // has no db default at the moment, so must be set
+func (in *NewItemRequest) itemData(itemID int64, itemMap map[string]interface{}) map[string]interface{} {
+	itemMap["ID"] = itemID
+	itemMap["idDefaultLanguage"] = in.LanguageID
+	if _, ok := itemMap["sContestPhase"]; !ok {
+		itemMap["sContestPhase"] = "Running"
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/items/add_item.go]: removed it

File: app/api/items/add_item.go [hunk 8]
```diff
@@ -3,145 +3,292 @@ package items
...
-	return &database.ItemString{
-		ID:          *types.NewInt64(id),
-		ItemID:      in.ID.Int64,
-		LanguageID:  in.Strings[0].LanguageID.Int64,
-		Title:       in.Strings[0].Title.String,
-		ImageURL:    in.Strings[0].ImageURL.String,
-		Subtitle:    in.Strings[0].Subtitle.String,
-		Description: in.Strings[0].Description.String,
+func (in *NewItemRequest) itemStringData(itemStringID, itemID int64, stringMap map[string]interface{}) map[string]interface{} {
+	stringMap["ID"] = itemStringID
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/items/add_item.go]: removed such methods

File: app/api/items/add_item.feature [hunk 9]
```diff
@@ -2,74 +2,209 @@ Feature: Add item
...
+    | 5577006791947779410 | Course | null |                 3 |              0 |        0 | null    | 1                | 0              | 0                       | 1        | 0         | default     | 0               | 0           | 0             | 0           | All             | null           | null           | 100             | null      | 0              | null          | 0               | 0            | null                 | null      | null                 | 0              | Running       | null   | 0        | 0              |
   And the table "items_strings" should be:
-    |                  ID | idItem  | idLanguage |   sTitle |          sImageUrl | sSubtitle |                 sDescription |
-    | 8674665223082153551 |      2  |          3 | my title | http://bit.ly/1234 | hard task | the goal of this task is ... |
+    |                  ID | idItem              | idLanguage |   sTitle |          sImageUrl | sSubtitle |                 sDescription |
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/items/add_item.feature]: fixed

File: app/api/items/add_item.go [hunk 10]
```diff
@@ -3,145 +3,292 @@ package items
...
 	}
 }
-func (in *NewItemRequest) itemItemData(id int64) *database.ItemItem {
-	return &database.ItemItem{
-		ID:           *types.NewInt64(id),
-		ChildItemID:  in.ID.Int64,
-		Order:        in.Parents[0].Order.Int32,
-		ParentItemID: in.Parents[0].ID.Int64,
+
+func (in *NewItemRequest) checkItemsRelationsCycles(store *database.DataStore) bool {
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/items/add_item.go]: renamed the method

File: app/api/items/add_item.go [hunk 11]
```diff
@@ -3,145 +3,292 @@ package items
...
-	}
-	service.MustNotBeError(err)
-	if !found {
-		return service.ErrForbidden(errors.New("cannot find the parent item"))
-	}
-	if !hasAccess {
-		return service.ErrForbidden(errors.New("insufficient access on the parent item"))
-	}
-	return service.NoError
+func constructTeamInGroupIDValidator(store *database.DataStore, user *database.User) validator.Func {
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/items/add_item.go]: added comments to validators

File: app/api/items/add_item.go [hunk 12]
```diff
@@ -3,145 +3,292 @@ package items
...
+	return validator.Func(func(fl validator.FieldLevel) bool {
+		var result int64
+		err := store.Groups().
+			OwnedBy(user).Where("groups.ID = ?", fl.Field().Interface().(int64)).
+			WithWriteLock().PluckFirst("1", &result).Error()
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/items/add_item.go]: Added the DB.HasRows() method

File: app/api/items/add_item.feature (hunk 1 above)
Comment (zenovich) [reply app/api/items/add_item.feature]: fixed

File: app/api/items/add_item.go (hunk 2 above)
Comment (zenovich) [reply app/api/items/add_item.go]: fixed

File: app/api/items/add_item.go (hunk 3 above)
Comment (zenovich) [reply app/api/items/add_item.go]: No, there is not. But I rewrote this part to use multi-insert

File: app/database/item_store.go (hunk 5 above)
Comment (zenovich) [reply app/database/item_store.go]: Fixed (although I'm not sure it became more readable with 'Count()')

File: app/api/items/add_item.go [hunk 13]
```diff
@@ -21,50 +21,52 @@ type itemChild struct {
...
-	} `json:"item" sql:"column:items"`
+	Item       item  `json:"item,squash"`
 	LanguageID int64 `json:"language_id" validate:"required,language_id"`
 	String     struct {
 		Title       string  `json:"title" validate:"required" sql:"column:sTitle"`
 		ImageURL    *string `json:"image_url" sql:"column:sImageUrl"`
 		Subtitle    *string `json:"subtitle" sql:"column:sSubtitle"`
 		Description *string `json:"description" sql:"column:sDescription"`
-	} `json:"string"`
+	} `json:"string,squash"`
```
Comment (smadbe) [inline app/api/items/add_item.go]: the parser for openapi will love this tag ;-)

File: app/database/item_store.go [hunk 14]
```diff
@@ -69,7 +69,7 @@ func (s *ItemStore) HasManagerAccess(user *User, itemIDs ...int64) (found bool,
 	err = s.GroupItems().MatchingUserAncestors(user).
 		WithWriteLock().
 		Where("idItem IN (?) AND (bManagerAccess OR bOwnerAccess)", itemIDs).
-		PluckFirst("COUNT(DISTINCT idItem)", &count).Error()
+		Select("COUNT(DISTINCT idItem)").Count(&count).Error()
```
Comment (smadbe) [inline app/database/item_store.go]: What I meant with the count was something allowing `s.GroupItems().Where("idItem IN (?).Count("DISTINCT idItem")` directly. But maybe that's not a good idea.

File: app/api/items/add_item.go (hunk 13 above)
Comment (zenovich) [reply app/api/items/add_item.go]: :))))

File: app/database/item_store.go (hunk 14 above)
Comment (zenovich) [reply app/database/item_store.go]: Gorm cannot do that. See here https://github.com/jinzhu/gorm/issues/970

**Pull request merged**